src/interp/interp-wasi.cc

```diff
@@ -76,18 +120,247 @@ class WasiInstance {
+      trace_stream->Writef("path_unlink_file %d %s\n", fd, path);
+    }
+    results[0].i32_ = uvwasi_path_unlink_file(uvwasi, fd, path, path_len);
+    return Result::Ok;
+  }
+
+  Result fd_prestat_get(const Values& params,
+                        Values& results,
+                        Trap::Ptr* trap) {
+    uvwasi_fd_t fd = params[0].i32_;
+    if (trace_stream) {
+      trace_stream->Writef("fd_prestat_get %d\n", fd);
+    }
+    CHECK_RESULT(getMemPtr<__wasi_prestat_t>(params[1].i32_, 1, nullptr, trap));
+    uvwasi_prestat_t buf;
+    results[0].i32_ = uvwasi_fd_prestat_get(uvwasi, fd, &buf);
+    uvwasi_serdes_write_prestat_t(memory->UnsafeData(), params[1].i32_, &buf);
+    if (trace_stream) {
+      trace_stream->Writef("fd_prestat_get -> %d\n", results[0].i32_);
+    }
+    return Result::Ok;
+  }
+
+  Result fd_prestat_dir_name(const Values& params,
+                             Values& results,
+                             Trap::Ptr* trap) {
+    uvwasi_fd_t fd = params[0].i32_;
+    uint32_t path_ptr = params[1].i32_;
+    uvwasi_size_t path_len = params[2].i32_;
+    if (trace_stream) {
+      trace_stream->Writef("fd_prestat_dir_name %d %d %d\n", fd, path_ptr,
+                           path_len);
+    }
+    char* path;
+    CHECK_RESULT(getMemPtr<char>(path_ptr, path_len, &path, trap));
+    results[0].i32_ = uvwasi_fd_prestat_dir_name(uvwasi, fd, path, path_len);
+    if (trace_stream) {
+      trace_stream->Writef("fd_prestat_dir_name %d -> %d %s %d\n", fd,
+                           results[0].i32_, path, path_len);
+    }
+    return Result::Ok;
+  }
+
+  Result fd_filestat_get(const Values& params,
+                         Values& results,
+                         Trap::Ptr* trap) {
+    /* __wasi_fd_filestat_get(__wasi_fd_t f, __wasi_filestat_t *buf) */
+    uvwasi_fd_t fd = params[0].i32_;
+    uint32_t filestat_ptr = params[1].i32_;
+    uvwasi_filestat_t buf;
+    results[0].i32_ = uvwasi_fd_filestat_get(uvwasi, fd, &buf);
+    uvwasi_serdes_write_filestat_t(memory->UnsafeData(), filestat_ptr, &buf);
+    if (trace_stream) {
+      trace_stream->Writef("fd_filestat_get -> size=%lu %d\n", buf.st_size,
+                           results[0].i32_);
+    }
+    return Result::Ok;
+  }
+
+  Result fd_fdstat_set_flags(const Values& params,
+                             Values& results,
+                             Trap::Ptr* trap) {
+    assert(false);
+    return Result::Ok;
+  }
+
   Result fd_fdstat_get(const Values& params, Values& results, Trap::Ptr* trap) {
     int32_t fd = params[0].i32_;
+    uint32_t stat_ptr = params[1].i32_;
     if (trace_stream) {
       trace_stream->Writef("fd_fdstat_get %d\n", fd);
     }
-    results[0].i32_ = __WASI_ERRNO_NOENT;
+    CHECK_RESULT(getMemPtr<__wasi_fdstat_t>(stat_ptr, 1, nullptr, trap));
+    uvwasi_fdstat_t statbuf;
+    results[0].i32_ = uvwasi_fd_fdstat_get(uvwasi, fd, &statbuf);
+    uvwasi_serdes_write_fdstat_t(memory->UnsafeData(), params[1].i32_,
```

Comment:
use `stat_ptr` instead of `params[1].i32_`?

Reply:
Done